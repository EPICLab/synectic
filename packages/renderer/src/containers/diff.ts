import {diffLines} from 'diff';

/**
 * Diffs two blocks of text, comparing line by line, based on the algorithm proposed
 * in "An O(ND) Difference Algorithm and its Variations" (Myers, 1986). Results are formatted
 * according to the unified diff patch used in the Git protocol, where +/- indicate lines
 * added or removed in order to transform from original to updated.
 * @param original The base version of code or text to evaluate.
 * @param updated The updated version of code or text to compare against original.
 * @returns {string} A diff patch that conforms to the unified diff patch format.
 */
export const diff = (original: string, updated: string): string => {
  const diffLineChanges = diffLines(original, updated);
  let readableDiff = '';

  diffLineChanges.map(segment => {
    const segmentLines = segment.value.split(/\n/g, segment.count);
    if (segment.added) {
      segmentLines.map(line => (readableDiff += `+${line}\n`));
    } else if (segment.removed) {
      segmentLines.map(line => (readableDiff += `-${line}\n`));
    } else {
      segmentLines.map(line => (readableDiff += ` ${line}\n`));
    }
  });

  return readableDiff;
};
