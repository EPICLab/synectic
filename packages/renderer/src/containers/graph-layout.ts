import type {UUID} from '@syn-types/app';
import dagre from 'dagre';
import type {Edge, Node} from 'reactflow';
import {MarkerType, isEdge, isNode} from 'reactflow';
import type {CommitGraph, CommitVertex} from './hooks/useGitGraph';

const nodesep = 80;

const layoutGraph = (graph: CommitGraph, topological: UUID[]): (Edge | Node)[] => {
  const elements: (Edge | Node)[] = [];
  for (const oid of topological) {
    const commit = graph.get(oid.toString());
    if (commit) {
      const node: Node = getNode(commit);
      const edges: Edge[] = getEdges(commit);
      elements.push(node);
      edges.map(edge => elements.push(edge));
    }
  }
  return layoutOptimizer(elements);
};

const layoutOptimizer = (rfGraph: (Edge | Node)[]): (Edge | Node)[] => {
  const graph = new dagre.graphlib.Graph();
  graph.setGraph({nodesep: nodesep});
  graph.setDefaultEdgeLabel(() => {
    return {};
  });

  rfGraph.filter(isNode).map(node => graph.setNode(node.id, {width: 5, height: 5}));
  rfGraph.filter(isEdge).map(edge => graph.setEdge(edge.source, edge.target));
  dagre.layout(graph);

  return rfGraph.map(els => {
    if (isNode(els)) {
      const graphNode = graph.node(els.id);
      return {...els, position: {x: graphNode.x, y: graphNode.y}};
    }
    return els;
  });
};

const getNode = (commit: CommitVertex): Node<CommitVertex> => {
  return {
    id: commit.oid.toString(),
    type: 'gitNode',
    data: commit,
    position: {x: 0, y: 0},
  };
};

const getEdges = (commit: CommitVertex): Edge[] => {
  return commit.parents
    .filter(parent => parent.length > 0)
    .map(parent => ({
      id: `e${parent.toString().slice(0, 7)}=${
        commit.staged ? commit.oid.toString() : commit.oid.toString().slice(0, 7)
      }`,
      source: parent.toString(),
      target: commit.oid.toString(),
      animated: commit.staged ? true : false,
      arrowHeadType: MarkerType.ArrowClosed,
    }));
};

export default layoutGraph;
