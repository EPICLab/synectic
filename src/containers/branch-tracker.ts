import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import { RootState } from '../store/root';
import { loadCard } from './handlers';
import { getMetafile } from './metafiles';

export const loadBranchVersions = (): ThunkAction<Promise<void>, RootState, undefined, Action> => {
  return async (dispatch) => {
    const metafile = await dispatch(getMetafile({ virtual: { name: 'Version Tracker', handler: 'Tracker' } }));
    if (metafile) dispatch(loadCard({ metafile: metafile }));
  };
}