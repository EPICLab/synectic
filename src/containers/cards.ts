import { DateTime } from 'luxon';
import { v4 } from 'uuid';

import type { Card } from '../types';
import { ActionKeys, NarrowActionType } from '../store/actions';
import { HandlerRequiredMetafile } from './handlers';

type AddCardAction = NarrowActionType<ActionKeys.ADD_CARD>;
type UpdateCardAction = NarrowActionType<ActionKeys.UPDATE_CARD>;

/**
 * Action Creator for composing a valid ADD_CARD Redux Action.
 * @param metafile The related `Metafile` object containing a valid `handler` field.
 * @return An `AddCardAction` object that can be dispatched via Redux.
 */
export const addCard = (metafile: HandlerRequiredMetafile): AddCardAction => {
  const card: Card = {
    id: v4(),
    name: metafile.name,
    created: DateTime.local(),
    modified: metafile.modified,
    left: 10,
    top: 25,
    type: metafile.handler,
    metafile: metafile.id
  };
  return {
    type: ActionKeys.ADD_CARD,
    id: card.id,
    card: card
  };
}

/**
 * Action Creator for composing a valid UPDATE_CARD Redux Action. If the current Redux store does not contain a
 * matching card (based on UUID) for the passed parameter, then dispatching this action will not result in any
 * changes in the Redux store state.
 * @param card A `Card` object containing new field values to be updated.
 * @return An `UpdateCardAction` object that can be dispatched via Redux, including an updated timestamp in the
 * `modified` field.
 */
export const updateCard = (card: Card): UpdateCardAction => {
  return {
    type: ActionKeys.UPDATE_CARD,
    id: card.id,
    card: { ...card, modified: DateTime.local() }
  }
}