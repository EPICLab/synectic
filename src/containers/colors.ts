export type ColorSet = {
  primary: string;
  secondary: string;
};

// colors from http://clrs.cc/
export const colorSets: { [color: string]: ColorSet } = {
  navy: { primary: '#001f3f', secondary: '#80b5ff' },
  blue: { primary: '#0074d9', secondary: '#b3dbff' },
  aqua: { primary: '#7fdbff', secondary: '#004966' },
  teal: { primary: '#39cccc', secondary: '#000000' },
  olive: { primary: '#3d9970', secondary: '#163728' },
  green: { primary: '#2ecc40', secondary: '#0e3e14' },
  lime: { primary: '#01ff70', secondary: '#00662c' },
  yellow: { primary: '#ffdc00', secondary: '#665800' },
  orange: { primary: '#ff851b', secondary: '#663000' },
  red: { primary: '#ff4136', secondary: '#800600' },
  maroon: { primary: '#85144b', secondary: '#eb7ab1' },
  fuchsia: { primary: '#f012be', secondary: '#65064f' },
  purple: { primary: '#b10dc9', secondary: '#efa9f9' },
  black: { primary: '#111111', secondary: '#dddddd' },
  gray: { primary: '#aaaaaa', secondary: '#000000' },
  silver: { primary: '#dddddd', secondary: '#000000' },
  white: { primary: '#ffffff', secondary: '#444444' },
  conflictRed: { primary: '#da6473', secondary: '' },
  stagedBlue: { primary: '#61aeee', secondary: '' },
  unstagedOrange: { primary: '#d19a66', secondary: '' },
  nominalGreen: { primary: '#50c878', secondary: '' }
};

export const getColor = (color: keyof typeof colorSets): ColorSet => {
  return colorSets[color];
};
