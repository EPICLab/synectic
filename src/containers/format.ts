import { flattenObject } from './flatten';

/** Requires properties to be defined, or excluded from the type otherwise.
 * Inspired by: https://stackoverflow.com/a/60574436
 */
export type NoUndefinedField<T> = {
  [P in keyof T]-?: Exclude<T[P], null | undefined | void>;
};

/** Requires at least one type property, similar to `Partial<T>` but excludes the empty object.
 * Inspired by: https://stackoverflow.com/questions/48230773/how-to-create-a-partial-like-that-requires-a-single-property-to-be-set/48244432#48244432
 */
export type AtLeastOne<T, U = { [K in keyof T]: Pick<T, K>; }> =
  Partial<T> & U[keyof U];

/** Requires exactly one type property in an object, no more and no less.
 * Inspired by: https://github.com/sindresorhus/type-fest/blob/main/source/require-exactly-one.d.ts
 */
export type ExactlyOne<T, U extends keyof T = keyof T> =
  { [K in U]: (Required<Pick<T, K>> & Partial<Record<Exclude<U, K>, never>>) }[U] & Omit<T, U>;

/**
 * Converts a JavaScript Object Notation (JSON) string into a typed object.
 * @param json A valid JSON string.
 * @return A typed object (or nested array of objects).
 */
export const deserialize = <T>(json: string): T => JSON.parse(json) as T;

/**
 * Filters an array and removes any undefined elements contained within it.
 * @param array The given array of elements that should be filtered for undefined.
 * @return The resulting array devoid of any undefined elements.
 */
export const removeUndefined = <T>(array: (T | undefined)[]): T[] => {
  return array.filter((item): item is T => typeof item !== 'undefined');
};

/**
 * Generic for deduplicating array of elements given a comparator function that
 * indicates whether a pair of elements should be considered duplicates.
 * @param arr An array of elements.
 * @param comparator A comparator function that indicates true if elements a
 * and b are duplicate, and false otherwise.
 * @return The resulting array devoid of duplicate elements.
 */
export const removeDuplicates = <T>(arr: T[], comparator: (a: T, b: T) => boolean): T[] => {
  return arr.reduce((accumulator: T[], current) => {
    if (!accumulator.some(item => comparator(item, current))) accumulator.push(current);
    return accumulator;
  }, []);
};

/**
 * Generic for asynchronously filtering an array of elements given an async predicate
 * function that resolves to indicate whether an element should be considered for
 * inclusion in the resulting array.
 * @param arr An array of elements.
 * @param predicate A predicate function that resolves to true if element `e` meets
 * the filter inclusion requirements, and false otherwise.
 * @return The resulting array devoid of elements that did not meet the predicate.
 */
export const asyncFilter = async <T>(arr: T[], predicate: (e: T) => Promise<boolean>): Promise<T[]> => {
  const results = await Promise.all(arr.map(predicate));
  return arr.filter((_v, index) => results[index]);
};

/**
 * Generic for filtering an object given a set of filtering keys for inclusion in the resulting object.
 * @param obj An object containing key-value indexed fields.
 * @param filter An array of key strings that indicate which key-value pairs should included.
 * @return The resulting object devoid of key-value fields that did not match the key filter.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const filterObject = <T extends { [key: string]: any }>(obj: T, filter: string[]): T[] => {
  const flattenedObj = flattenObject(obj);
  return Object.entries(flattenedObj).filter(([key, _]) => filter.includes(key)).map(([_, value]) => value);
}

/**
 * Expand an array of path elements (i.e. a dot-format path such as `user.name`) into a successive series of named
 * objects that contain any subsequent path elements and/or the updated value.
 * @param path An array of path elements ordered in sequence (i.e. `user.name.lastName` would be `[user, name, lastName]`).
 * @param value The value to insert or update at the specific path in the return object.
 * @return An object containing a key-value structure of embedded container objects according to the given path.
 */
export const objectifyPath = (path: Array<string | number>, value: string | boolean | number | undefined):
  Record<string | number, unknown> => {
  return path
    .slice(0, -1)
    .reduceRight((prev: Record<string | number, unknown>, curr) => { return { [curr]: prev } }, { [path[path.length - 1]]: value });
};

/**
 * Compares two `ArrayBufferLike` objects for equality; compatible objects include `TypedArray`, `DataView`, 
 * and `Buffer`. However, `Buffer` has `.equals()` and `.compare()` functions for comparisons and these should 
 * be used in lieu of this function. `TypedArray` is a class of objects, which includes all of the following: 
 * `Int8Array`, `Uint8Array`, `Uint8ClampedArray`, `Int16Array` ,`Int32Array`, `Uint32Array`, `Float32Array`, 
 * `Float64Array`, `BigInt64Array`, `BigUint64Array`.
 * @param buf1 An `ArrayBufferLike` object.
 * @param buf2 An `ArrayBufferLike` object.
 * @return A boolean indicating true if the `ArrayBufferLike` objects are equal, or false otherwise.
 */
export const equalArrayBuffers = (buf1: ArrayBufferLike, buf2: ArrayBufferLike): boolean => {
  if (buf1.byteLength != buf2.byteLength) return false;
  const dv1 = new Int8Array(buf1);
  const dv2 = new Int8Array(buf2);
  for (let i = 0; i != buf1.byteLength; i++) {
    if (dv1[i] != dv2[i]) return false;
  }
  return true;
}

/**
 * Convert a `Buffer` object into an `ArrayBuffer` object.
 * @param buf A `Buffer` object representing binary data in the form of a sequence of bytes.
 * @return A new `ArrayBuffer` object containing a copy of all data in `Buffer` object.
 */
export const toArrayBuffer = (buf: Buffer): ArrayBuffer => {
  const ab = new ArrayBuffer(buf.length);
  const view = new Uint8Array(ab);
  for (let i = 0; i < buf.length; i++) {
    view[i] = buf[i];
  }
  return ab;
};