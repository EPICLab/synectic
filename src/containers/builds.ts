import { exec } from 'child_process';
import util from 'util';
import { checkout } from 'isomorphic-git';
import * as fs from 'fs-extra';
import { join } from 'path';

import type { Repository } from '../types';
import { clone, merge } from '../containers/git-porcelain';
import { readDirAsync } from './io';
import { removeUndefinedProperties } from './format';

const promiseExec = util.promisify(exec);

export const tempClone = async (repo: Repository, branch: string): Promise<string> => {
  const cloneRoot = join(repo.root.toString(), '../', '.syn');
  await clone({ repo: repo, dir: cloneRoot, ref: branch });
  return cloneRoot;
}

export const build = async (repo: Repository, base: string, compare: string): Promise<{ installCode: number; buildCode: number; }> => {
  const cloneRoot = await tempClone(repo, base);
  await checkout({ fs: fs, dir: cloneRoot, ref: base });

  const mergeResult = await merge(cloneRoot, base, compare);
  const ref = removeUndefinedProperties({ ref: mergeResult.oid });
  // this next step seems unnecessary, but isomorphic-git.merge stages a reversal changeset that negates the results of the merge
  // reverting the target branch back to the merged commit (via checkout) allows these reversal changesets to be removed
  await checkout({ fs: fs, dir: cloneRoot, ...ref });

  const rootFiles = await readDirAsync(cloneRoot);
  const packageManager = rootFiles.find(file => file === 'yarn.lock') ? 'yarn' : 'npm';

  let [installCode, buildCode] = [-1, -1];
  try {
    const installResults = promiseExec(`${packageManager} install`, { cwd: cloneRoot });
    installResults.child.stdout?.on('data', data => console.log('INSTALL: ' + data));
    installResults.child.stderr?.on('data', data => console.log('INSTALL error: ' + data));
    installResults.child.on('close', code => {
      console.log(`INSTALL 'close' listener found code: ${code}`);
      (code ? (installCode = code) : null);
    });
    installResults.child.on('exit', code => {
      console.log(`INSTALL 'exit' listener found code: ${code}`);
      (code ? (installCode = code) : null);
    });
    installResults.child.on('error', error => {
      console.log('INSTALL \'error\' listener found error:');
      console.log({ error });
    });
    await installResults;
  } catch (e) {
    console.log('INSTALL ERROR');
    console.log(e);
  }

  if (installCode === 0) {
    const packageManagerBuildScript = packageManager === 'yarn' ? 'run' : 'run-script';
    try {
      const buildResults = promiseExec(`${packageManager} ${packageManagerBuildScript} build`, { cwd: cloneRoot });
      buildResults.child.stdout?.on('data', data => console.log('BUILD: ' + data));
      buildResults.child.stderr?.on('data', data => console.log('BUILD error: ' + data));
      buildResults.child.on('close', code => {
        console.log(`BUILD 'close' listener found code: ${code}`);
        (code ? (buildCode = code) : null);
      });
      buildResults.child.on('exit', code => {
        console.log(`BUILD 'exit' listener found code: ${code}`);
        (code ? (buildCode = code) : null);
      });
      buildResults.child.on('error', error => {
        console.log('BUILD \'error\' listener found error:');
        console.log({ error });
      });
      await buildResults;
    } catch (e) {
      console.log('BUILD ERROR');
      console.error(e);
    }
  }

  fs.remove(cloneRoot, (error) => console.log(error));

  return { installCode: installCode, buildCode: buildCode };
}