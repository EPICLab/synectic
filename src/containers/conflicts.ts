import * as path from 'path';
import { PathLike } from 'fs-extra';
import * as io from './io';
import { getIgnore } from './git-plumbing';
import { asyncFilter, removeUndefined } from './format';
import { Ignore } from 'ignore';
import { createAsyncThunk } from '@reduxjs/toolkit';
import { AppThunkAPI } from '../store/hooks';
import { v4 } from 'uuid';
import { fetchMetafile } from '../store/thunks/metafiles';
import { DateTime } from 'luxon';
import { loadCard } from '../store/thunks/handlers';
import { getBranchRoot, getRoot, getWorktreePaths } from './git-path';

export type Conflict = {
    /** The relative or absolute path to the file containing conflicts. */
    filepath: PathLike,
    /** An array of indexed conflicts indicating the starting position of each conflict in the file. */
    conflicts: number[]
}

export const loadConflictManagers = createAsyncThunk<void, void, AppThunkAPI>(
    'conflicts/loadConflictManagers',
    async (_, thunkAPI) => {
        const repos = removeUndefined(Object.values(thunkAPI.getState().repos.entities));
        await Promise.all(repos.map(async repo => {
            await Promise.all(repo.local.map(async branchId => {
                const branch = thunkAPI.getState().branches.entities[branchId];
                const root = branch ? await getBranchRoot(repo.root, branch.ref) : undefined;
                const conflicts = await checkProject(root);
                if (branch && root && conflicts.length > 0) {
                    const { base, compare } = await resolveConflictBranches(root);
                    const conflictManager = await thunkAPI.dispatch(fetchMetafile({
                        virtual: {
                            id: v4(),
                            modified: DateTime.local().valueOf(),
                            name: `Conflicts`,
                            handler: 'ConflictManager',
                            repo: repo.id,
                            path: root,
                            merging: { base: (base ? base : branch.ref), compare: compare }
                        }
                    })).unwrap();
                    await thunkAPI.dispatch(loadCard({ metafile: conflictManager }));
                }
            }));
        }));
    }
);

export const resolveConflictBranches = async (root: PathLike): Promise<{ base: string | undefined, compare: string }> => {
    const branchPattern = /(?<=^Merge branch(es)? .*)('.+?')+/gm;
    const worktree = await getWorktreePaths(root);
    const mergeMsg = worktree.gitdir ? await io.readFileAsync(path.join(worktree.gitdir.toString(), 'MERGE_MSG'), { encoding: 'utf-8' }) : '';
    const match = mergeMsg.match(branchPattern);
    return match
        ? match.length === 2
            ? { base: match[0].replace(/['"]+/g, ''), compare: match[1].replace(/['"]+/g, '') }
            : { base: undefined, compare: match[0].replace(/['"]+/g, '') }
        : { base: undefined, compare: '' };
}

export const checkFilepath = async (filepath: PathLike, ignoreManager?: Ignore): Promise<Conflict | undefined> => {
    const root = await getRoot(filepath);
    if (root && !(await io.isDirectory(filepath))) {
        const conflictPattern = /<<<<<<<[^]+?=======[^]+?>>>>>>>/gm;
        const ignore = ignoreManager ? ignoreManager : (await getIgnore(root, true));
        if (ignore.ignores(path.relative(root.toString(), filepath.toString()))) return undefined;
        const content = await io.readFileAsync(filepath, { encoding: 'utf-8' });
        const matches = Array.from(content.matchAll(conflictPattern));
        const matching = { filepath: filepath, conflicts: removeUndefined(matches.map(m => m.index)) };
        if (matching.conflicts.length > 0) return matching;
    }
    return undefined;
}

export const checkProject = async (root: PathLike | undefined): Promise<Conflict[]> => {
    if (root) {
        const conflictPattern = /<<<<<<<[^]+?=======[^]+?>>>>>>>/gm;
        const ignore = (await getIgnore(root, true));

        const paths = (await io.readDirAsyncDepth(root, 1))
            .filter(p => p !== root)                                          // filter root filepath from results
            .filter(p => !ignore.ignores(path.relative(root.toString(), p))); // filter based on git-ignore rules
        const dirpaths = await asyncFilter(paths, async p => io.isDirectory(p));
        const filepaths = paths.filter(p => !dirpaths.includes(p));

        const matching = await Promise.all(filepaths.map(async f => {
            const content = await io.readFileAsync(f, { encoding: 'utf-8' });
            const matches = Array.from(content.matchAll(conflictPattern));
            return { filepath: f, conflicts: removeUndefined(matches.map(m => m.index)) };
        }));

        return matching.filter(match => match.conflicts.length > 0);
    } else {
        return [];
    }
}