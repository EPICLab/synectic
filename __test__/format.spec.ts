import * as format from '../src/containers/format';

describe('format.deserialize', () => {
  it('deserialize to parse a JSON string into a TypeScript object', () => {
    const json = '{"result":true, "count":42}';
    type typedJson = { result: boolean; count: number };
    const deserializedJson = format.deserialize<typedJson>(json);
    expect(typeof deserializedJson.result).toBe('boolean');
    expect(deserializedJson.result).toBe(true);
    expect(deserializedJson).toMatchSnapshot();
  });

  it('deserialize fails with an error on malformed JSON', () => {
    // eslint-disable-next-line no-useless-escape
    const malformedJson = '{ "key": "Something \\\\"Name\\\\" something\", "anotherkey": "value" }';
    expect(() => format.deserialize(malformedJson)).toThrow(SyntaxError);
  });
});

describe('format.equalArrayBuffers', () => {
  it('equalArrayBuffers identifies equal and non-equal Buffers containing UTF-8 string data', () => {
    const buf1 = Buffer.from('test content');
    const buf2 = Buffer.from('test content');
    const buf3 = Buffer.from('other content');
    expect(format.equalArrayBuffers(buf1, buf2)).toBe(true);
    expect(format.equalArrayBuffers(buf2, buf3)).toBe(false);
  });

  it('equalArrayBuffers identifies equal and non-equal Buffers containing binary data', () => {
    const buf1 = Buffer.from([23, 13, 99]);
    const buf2 = Buffer.from([23, 13, 99]);
    const buf3 = Buffer.from([14, 54, 116, 34, 28]);
    expect(format.equalArrayBuffers(buf1, buf2)).toBe(true);
    expect(format.equalArrayBuffers(buf3, buf2)).toBe(false);
  });

  it('equalArrayBuffers identifies equal and non-equal Uint8Arrays containing binary data', () => {
    const arr1 = new Uint8Array([23, 13, 99]);
    const arr2 = new Uint8Array([23, 13, 99]);
    const arr3 = new Uint8Array([14, 54, 116, 34, 28]);
    const arr4 = new Uint8Array([14, 53, 116, 34, 28]);
    expect(format.equalArrayBuffers(arr1, arr2)).toBe(true);
    expect(format.equalArrayBuffers(arr3, arr2)).toBe(false);
    expect(format.equalArrayBuffers(arr3, arr4)).toBe(false);
  });

  it('equalArrayBuffers identifies equal and non-equal ArrayBufferLike objects of different types', () => {
    const buf1 = Buffer.from([23, 13, 99]);
    const buf2 = Buffer.from([14, 54, 116, 34, 28]);
    const arr1 = new Uint8Array([23, 13, 99]);
    const arr2 = new Uint8Array([14, 53, 116, 34, 28]);
    expect(format.equalArrayBuffers(buf1, arr1)).toBe(true);
    expect(format.equalArrayBuffers(buf2, arr2)).toBe(false);
  });
});

describe('format.toArrayBuffer', () => {
  it('toArrayBuffer converts Buffer containing UTF-8 string data', () => {
    const buf1 = Buffer.from('test content');
    const arr1 = format.toArrayBuffer(buf1);
    expect(format.equalArrayBuffers(buf1, arr1)).toBe(true);
  });

  it('toArrayBuffer converts Buffer containing binary data', () => {
    const buf1 = Buffer.from([14, 54, 116, 34, 28]);
    const arr1 = format.toArrayBuffer(buf1);
    expect(format.equalArrayBuffers(buf1, arr1)).toBe(true);
  });

  it('toArrayBuffer converts Uint8Array containing binary data', () => {
    const arr1 = new Uint8Array([14, 54, 116, 34, 28]);
    const buf1 = Buffer.from(arr1);
    const arr2 = format.toArrayBuffer(buf1);
    expect(format.equalArrayBuffers(arr1, arr2)).toBe(true);
  });
});