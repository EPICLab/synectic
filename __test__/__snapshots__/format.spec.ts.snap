// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/format deserialize to parse a JSON string into a TypeScript object 1`] = `
Object {
  "count": 42,
  "result": true,
}
`;
